extern/alpaka/cmake/alpaka-functions.cmake

```diff
@@ -0,0 +1,127 @@
+# which is setup to use the correct compiler flags depending on the build type.
+#
+# Usage: traccc_add_alpaka_library( traccc_core core
+#                                   [TYPE SHARED/INTERFACE/STATIC]
+#                                   include/source1.hpp source2.cpp )
+#
+function( traccc_add_alpaka_library fullname basename )
+
+   # Parse the function's options.
+   cmake_parse_arguments( ARG "" "TYPE" "" ${ARGN} )
+
+   # Decide what sources to give to the library.
+   set( _sources ${ARG_UNPARSED_ARGUMENTS} )
+   if( "${ARG_TYPE}" STREQUAL "INTERFACE" )
+      set( _sources )
+   endif()
+
+   # Create the library.
+   alpaka_add_library( ${fullname} ${ARG_TYPE} ${_sources} )
+
+   # Set up how clients should find its headers.
+   if( IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/include/" )
+      set( _depType PUBLIC )
+      if( "${ARG_TYPE}" STREQUAL "INTERFACE" )
+         set( _depType INTERFACE )
+      endif()
+      target_include_directories( ${fullname} ${_depType}
+         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+         $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}> )
+      unset( _depType )
+   endif()
+
+   # Make sure that the library is available as "traccc::${basename}" in every
+   # situation.
+   set_target_properties( ${fullname} PROPERTIES EXPORT_NAME ${basename} )
+   add_library( traccc::${basename} ALIAS ${fullname} )
+
+   # Specify the (SO)VERSION of the library.
+   if( NOT "${ARG_TYPE}" STREQUAL "INTERFACE" )
+      set_target_properties( ${fullname} PROPERTIES
+         VERSION ${PROJECT_VERSION}
+         SOVERSION ${PROJECT_VERSION_MAJOR} )
+   endif()
+
+   # Set up the installation of the library and its headers.
+   install( TARGETS ${fullname}
+      EXPORT traccc-exports
+      LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+      ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+      RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}" )
+   if( IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/include/" )
+      install( DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/include/"
+         DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}" )
+   endif()
+
+endfunction( traccc_add_alpaka_library )
+
+# Function for declaring the (installable) executables of the project
+# This version calls the alpaka_add_executable() function to create the executable,
+# which is setup to use the correct compiler flags depending on the build type.
+#
+# Usage: traccc_add_alpaka_executable( super_exe super_source.cpp )
+#
+function( traccc_add_alpaka_executable name )
+
+   # Parse the function's options.
+   cmake_parse_arguments( ARG "" "" "LINK_LIBRARIES" ${ARGN} )
+
+   # Set up the executable.
+   set( exe_name "traccc_${name}" )
+   add_executable( ${exe_name} ${ARG_UNPARSED_ARGUMENTS} )
+   if( ARG_LINK_LIBRARIES )
+      target_link_libraries( ${exe_name} PRIVATE ${ARG_LINK_LIBRARIES} )
+   endif()
+
+   target_compile_definitions(${exe_name} PRIVATE ALPAKA_HOST_ONLY)
```

Comment:
This is the key line in this function.

Without it, the example code will go via whatever compiler makes sense for the current accelerator (i.e. it will go via `nvcc` when you are using CUDA).

That isn't what we want: the examples (and I think all executables, but it is something to consider going forward), should be compiled as regular CXX files, and the device code should all live in the relevant `traccc::XXX` library.

I somehow got away with that in the old code, and this code until recently, but it looks like the recent detray inclusion (and all the templating code in the detray functions) causes `nvcc` to fall over.

This function makes alpaka fine to compile as a normal C++ executable, whilst all the actual library code that we want accelerated is compiled with `nvcc`, as expected, over in `traccc::alpaka`.

(Without it Alpaka will complain that you are using a C++ compiler, and not `nvcc` etc. This is because, well I think because, alpaka is a header only lib, so you end up pulling in all the alpaka headers very quickly, then you hit warnings over in those headers due to the combination of the current compiler + current enabled backend. With this flag, the warnings are not enabled, and the code builds as you'd expect).

Reply:
This then means that `traccc::alpaka` exposes Alpaka in its public interface. :frowning: We need to fix that. Neither `traccc::cuda`, nor `traccc::sycl` expose the clients of those libraries to CUDA and SYCL headers respectively. :thinking: `traccc::alpaka` must also be able to provide an interface that doesn't include any Alpaka headers...